### PR TITLE
Added RouteRetrievalEvent

### PR DESCRIPTION
**Pull request by devotaaabel**
- [x]  Create Event object
- [x]  Create and call the event when routes are retrieved/deserialized
- [x]  Add SessionId as required by the schema
- [x] Add other important attributes
- [x] Add tests

Looking for some feedback on where the event is created and sent.

### CODE REVIEW

File: libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/navigation/NavigationTelemetryUtils.java [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+
+import java.text.SimpleDateFormat;
+import java.util.Date;
+import java.util.Locale;
+
+class NavigationTelemetryUtils {
+  private static final String DATE_AND_TIME_PATTERN = "yyyy-MM-dd'T'HH:mm:ss.SSSZ";
+  private static final SimpleDateFormat dateFormat = new SimpleDateFormat(DATE_AND_TIME_PATTERN, Locale.US);
+
+  static String obtainCurrentDate() {
```
Comment (Guardiola31337) [inline libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/navigation/NavigationTelemetryUtils.java]: IMO, objects and methods should live / be where their context is. From my experience utility classes become a drawer easily and we should treat them carefully.

Comment (devotaaabel) [reply libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/navigation/NavigationTelemetryUtils.java]: @Guardiola31337 I copied this from a protected method in the telemetry library. Are you saying this should live in the event object? I imagine other navigation events will also need this method.

Comment (Guardiola31337) [reply libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/navigation/NavigationTelemetryUtils.java]: Yeah, I meant where it's called / obtained. 
> I imagine other navigation events will also need this method.

For now, that's not the case so I'd defer that decision to the _Last Responsible Moment_ / to the time in which the method is called from multiple classes.

In any case, that was a minor comment that I wanted to share.

**Comment by devotaaabel** [comment]
cc. @1ec5

**Comment by devotaaabel** [comment]
@danesfeder @Guardiola31337 This is ready for review. Waiting for feedback before fixing/adding tests

File: libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/navigation/Attribute.java [hunk 2]
```diff
@@ -0,0 +1,49 @@
+package com.mapbox.services.android.navigation.v5.navigation;
+
+import android.os.Parcel;
+import android.os.Parcelable;
+
+public class Attribute implements Parcelable {
```
Comment (danesfeder) [inline libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/navigation/Attribute.java]: This class can be package-private

File: libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/navigation/Counter.java [hunk 3]
```diff
@@ -0,0 +1,30 @@
+package com.mapbox.services.android.navigation.v5.navigation;
+
+import android.os.Parcelable;
+
+import com.google.gson.annotations.SerializedName;
+
+public abstract class Counter<N extends Number> implements Parcelable {
```
Comment (danesfeder) [inline libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/navigation/Counter.java]: can also be package-private

File: libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/navigation/DoubleCounter.java [hunk 4]
```diff
@@ -0,0 +1,32 @@
+package com.mapbox.services.android.navigation.v5.navigation;
+
+import android.os.Parcel;
+
+public class DoubleCounter extends Counter<Double> {
```
Comment (danesfeder) [inline libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/navigation/DoubleCounter.java]: can also be package-private

File: libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/navigation/IntCounter.java [hunk 5]
```diff
@@ -0,0 +1,32 @@
+package com.mapbox.services.android.navigation.v5.navigation;
+
+import android.os.Parcel;
+
+public class IntCounter extends Counter<Integer> {
```
Comment (danesfeder) [inline libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/navigation/IntCounter.java]: can also be package-private

File: libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/navigation/LongCounter.java [hunk 6]
```diff
@@ -0,0 +1,37 @@
+package com.mapbox.services.android.navigation.v5.navigation;
+
+import android.os.Parcel;
+
+public class LongCounter extends Counter<Long> {
```
Comment (danesfeder) [inline libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/navigation/LongCounter.java]: can also be package-private

File: libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/navigation/NavigationPerformanceEvent.java [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+
+import com.mapbox.android.telemetry.Event;
+
+import java.text.SimpleDateFormat;
+import java.util.ArrayList;
+import java.util.Date;
+import java.util.List;
+import java.util.Locale;
+
+public class NavigationPerformanceEvent extends Event implements Parcelable {
```
Comment (danesfeder) [inline libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/navigation/NavigationPerformanceEvent.java]: can also be package-private

File: libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/navigation/OfflineRouteRetrievalTask.java [hunk 8]
```diff
@@ -41,7 +53,15 @@ protected DirectionsRoute doInBackground(OfflineRoute... offlineRoutes) {
   @Override
   protected void onPostExecute(DirectionsRoute offlineRoute) {
     if (offlineRoute != null) {
+      routeRetrievalBuilder.end();
```
Comment (danesfeder) [inline libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/navigation/OfflineRouteRetrievalTask.java]: I'm wondering if `AsyncTask` here could further "delay" the timing here for offline route retrieval.  Maybe we should end in the `doInBackground`?  If it's negligible then no worries.

File: libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/navigation/NavigationRoute.java [hunk 9]
```diff
@@ -76,8 +77,26 @@ static Builder builder(Context context, LocaleUtils localeUtils) {
...
+  public void getRoute(final Callback<DirectionsResponse> callback) {
+    final RouteRetrievalInfo.Builder routeRetrievalBuilder = RouteRetrievalInfo.builder().start();
+    mapboxDirections.enqueueCall(new Callback<DirectionsResponse>() {
+      @Override
+      public void onResponse(Call<DirectionsResponse> call, Response<DirectionsResponse> response) {
+        routeRetrievalBuilder.end();
+        callback.onResponse(call, response);
+        NavigationTelemetry.getInstance().routeRetrievalEvent(
+          routeRetrievalBuilder
+            .route(response.body().routes().get(0))
```
Comment (danesfeder) [inline libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/navigation/NavigationRoute.java]: Does this route retrieval event only consider the first route returned in a directions API response?  If so, we should check the response is successful and has a valid list of route before getting `0` - possible `IndexOutOfBounds` exception.

File: libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/navigation/RouteRetrievalInfo.java [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+import com.mapbox.api.directions.v5.models.DirectionsRoute;
+import com.mapbox.api.directions.v5.models.LegStep;
+import com.mapbox.api.directions.v5.models.RouteLeg;
+import com.mapbox.api.directions.v5.models.RouteOptions;
+import com.mapbox.services.android.navigation.v5.exception.NavigationException;
+
+import java.util.List;
+
+@AutoValue
+abstract class RouteRetrievalInfo {
```
Comment (danesfeder) [inline libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/navigation/RouteRetrievalInfo.java]: We can add tests for the logic that occurs within `RouteRetrievalInfo`

**Comment by devotaaabel** [comment]
@danesfeder  I checked with debugging enabled that the event is being sent, is there more than I should verify?

**Comment by devotaaabel** [comment]
@danesfeder ready for re-review

File: libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/navigation/Attribute.java [hunk 11]
```diff
@@ -0,0 +1,41 @@
+package com.mapbox.services.android.navigation.v5.navigation;
+
+import android.os.Parcel;
+import android.os.Parcelable;
+
+class Attribute implements Parcelable {
+  private final String name;
```
Comment (andrlee) [inline libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/navigation/Attribute.java]: shouldn't these fields be tagged with `@SerializedName` annotations as well?

**Comment by devotaaabel** [comment]
@andrlee Thanks for the feedback! 

> Events are only Parcelable for backwards compatibility, so stubs are sufficient, internal implementation should not be driven by it.

I'll definitely keep this in mind and try to use it to simplify the implementation. As far as the other comments, I hope we can discuss this in more detail later today.

File: libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/navigation/NavigationPerformanceEvent.java [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+import java.util.Locale;
+
+@SuppressLint("ParcelCreator")
+@SuppressWarnings("ParcelableCreator")
+public class NavigationPerformanceEvent extends Event implements Parcelable {
+  private static final String DATE_AND_TIME_PATTERN = "yyyy-MM-dd'T'HH:mm:ss.SSSZ";
+  private static final SimpleDateFormat dateFormat = new SimpleDateFormat(DATE_AND_TIME_PATTERN, Locale.US);
+  private static final String PERFORMANCE_TRACE = "performance.trace";
+
+  protected final String event;
```
Comment (Guardiola31337) [inline libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/navigation/NavigationPerformanceEvent.java]: Do these properties need to be `protected`? 🤔

File: libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/navigation/NavigationPerformanceEvent.java [hunk 13]
```diff
@@ -0,0 +1,62 @@
...
+    this.counters = counters;
+    this.attributes = attributes;
+  }
+
+  NavigationPerformanceEvent(String sessionId) {
+    this(PERFORMANCE_TRACE, obtainCurrentDate(), sessionId,
+      new ArrayList<Counter>(), new ArrayList<Attribute>());
+  }
+
+  private static String obtainCurrentDate() {
```
Comment (Guardiola31337) [inline libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/navigation/NavigationPerformanceEvent.java]: I believe this doesn't need to be `static`.

File: libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/navigation/NavigationRoute.java [hunk 14]
```diff
@@ -76,8 +86,24 @@ static Builder builder(Context context, LocaleUtils localeUtils) {
...
-    mapboxDirections.enqueueCall(callback);
+  public void getRoute(final Callback<DirectionsResponse> callback) {
+    elapsedTimeBuilder.start();
+
+    mapboxDirections.enqueueCall(new Callback<DirectionsResponse>() {
+      @Override
+      public void onResponse(Call<DirectionsResponse> call, Response<DirectionsResponse> response) {
+        elapsedTimeBuilder.end();
+        callback.onResponse(call, response);
+        if (response.body().routes() != null && !response.body().routes().isEmpty()) {
```
Comment (danesfeder) [inline libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/navigation/NavigationRoute.java]: @devotaaabel should we remove this check?  Or do we only want to send the even if we have routes in the response?

Comment (devotaaabel) [reply libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/navigation/NavigationRoute.java]: Per slack convo, we're only going to send this event when there are routes in the request

File: libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/navigation/NavigationPerformanceEvent.java [hunk 15]
```diff
@@ -0,0 +1,56 @@
...
+
+  private String obtainCurrentDate() {
+    return dateFormat.format(new Date());
+  }
+
+  void addCounter(Counter counter) {
+    counters.add(counter);
+  }
+
+  void addAttribute(Attribute attribute) {
```
Comment (1ec5) [inline libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/navigation/NavigationPerformanceEvent.java]: Per chat, it looks like the call to this method (for adding the `routeUuid` attribute) fell out of the PR unintentionally.

**Comment by devotaaabel** [comment]
@Guardiola31337 @danesfeder before I test that these are showing up in mode, can you verify this looks ok? Specifically, I wanted to see if you all had a preference on whether the two pieces of information (`elapsedTime` and `routeUuid`) should be in an info object or are ok on their own.

**Comment by devotaaabel** [comment]
This was verified in mode (thanks @danesfeder), ready for final review

**Pull request merged**